Fixed an issue where restrictedToMinimumLevel from appsettings.json was ignored.

ashishdhingra opened the pull request:
*Issue #:* #167 

*Description of changes:*
Fixed an issue where restrictedToMinimumLevel from appsettings.json was ignored.

___
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

ashovlin reviewed (commented):
1. How have you tested this? 
2. We do have automated tests for SeriLog, may be nice to add this.
3. Would it be helpful to show to where to configure this in the `appsettings.json` in the SeriLog samples?

ashishdhingra commented:
> 1. How have you tested this?
> 2. We do have automated tests for SeriLog, may be nice to add this.
> 3. Would it be helpful to show to where to configure this in the `appsettings.json` in the SeriLog samples?

1. Tested locally.
2. Added automated test. Cannot re-use base class since the setting is specific to Serilog.
3. Added sample. Also corrected solution to include Serilog sample folder.

ashovlin commented on test/AWS.Logger.SeriLog.Tests/SeriLoggerTestClass.cs:
```diff
@@ -50,6 +51,55 @@ public void MultiThreadBufferFullTest()
+
+            Log.Logger = new LoggerConfiguration().
+                ReadFrom.Configuration(configuration).
+                WriteTo.AWSSeriLog(configuration, null, null, Serilog.Events.LogEventLevel.Error).CreateLogger();
```
I'm a little confused by the test design - if the issue this is addressing is for not reading the `restrictedToMinimumLevel` from `appsettings.json` - why are we specifying the minimum level here instead of in the json?

ashishdhingra replied:
@ashovlin Somehow while running through test runner, it doesn't follow the code path it is supposed to follow. While the actual application debugging works fine. You could try getting the branch and see what I mean. I tried it several times but then gave up. We can discuss when I'm back online.

ashishdhingra replied:
I have fixed the issue. The version of `Serilog.Settings.Configuration` needed to be updated (test suite was using the old version having the config issue). Also updated the version of `Serilog` to `2.10.0`.

ashishdhingra replied:
Based on @ppittle's suggestion, min-upgraded Serilog to version 2.6.0 (from 2.5.0) since it was anyway required by old and new (updated). version of Serilog.Settings.Configuration.

ppittle commented on src/AWS.Logger.SeriLog/AWS.Logger.SeriLog.csproj:
```diff
@@ -30,7 +30,7 @@
 
   <ItemGroup>
-    <PackageReference Include="Serilog" Version="2.5.0" />
+    <PackageReference Include="Serilog" Version="2.10.0" />
```
Does the fix require upgrading to `2.10.0` or can we stay on `2.5.0`?

ashishdhingra replied:
@ppittle Customer disn't ask, but there was similar ticket for Log4Net logger, and since we are upgrading `Serilog.Settings.Configuration` to latest version to fix test suite, I thought it would be good to upgrade Serilog reference as well.

ppittle replied:
I think we should take the lowest version dependency possible on Serilog.  That way if a customer is already using Serilog they wont need to worry about upgrading their dependency to use our library.

ashishdhingra replied:
I will revert the change to upgrade the Serilog version to minimize the scope of this PR, test it locally and push new changes.

ashishdhingra replied:
I have min-upgraded `Serilog` to version `2.6.0` (from `2.5.0`) since it was anyway required by old and new (updated). version of `Serilog.Settings.Configuration`.

ashovlin replied:
Agree with @ppittle's point where it's seems nice to stay at minimum if possible. I believe the Log4Net upgrade initially started due to a CVE for the version we were on. 

Could the test project depend on 2.6.0 if it's required there, then the source stays at 2.5.0?

ashishdhingra replied:
I would prefer to keep test project and source at same level, i.e., at `2.6.0` since the package `Serilog.Settings.Configuration` still had the issue in the previous version and the last good version is dependent on `Serilog 2.6.0`.

ppittle commented on test/AWS.Logger.SeriLog.Tests/AWSSeriLogGroupRestrictedToMinimumLevel.json:
```diff
@@ -0,0 +1,18 @@
+﻿{
+  "Serilog": {
+    "Using": [ "AWS.Logger.SeriLog" ],
+    "MinimumLevel": "Debug",
```
Why would `MinimumLevel` be **Debug** but `restrictedToMiniumLevel` be **Error**?  Is this a valid configuration case? Is the order of operations documented?

ashishdhingra replied:
`restrictedToMiniumLevel` overrides the behavior for a sink. This is a valid configuration. If you refer the test case, we get 6 log lines from CloudWatch (for Error), instead of 10 (which would have included Debug).
